design-proposals/klip-7.md

```diff
@@ -0,0 +1,176 @@
+    return selectExpressions;
+  }
+
+  @Override
+  public List<ExecutionStep> getSources() {
+    return Collections.singletonList(source);
+  }
+
+  @Override
+  public KStream build(final KStream input) {
+      final Selection selection = new Selection(selectExpressions);
+      return kstream.mapValues(selection.getSelectValueMapper()),
+  }
+}
+```
+
+I've written up a basic implementation of Execution Plans which includes all the Execution Steps we'd need, and can be found here (https://github.com/confluentinc/ksql/blob/ksql-execution-plans/ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java) (steps are under ksql-engine/src/main/java/io/confluent/ksql/structured/execution):
+
+#### Serializability
+The Execution Plan needs to be JSON-serializable. This way, we can persist the plan generated for each case in our Query Translation tests, and verify that it is kept the same across versions.
+
+#### Versioning
+We may need to version Execution Steps. Remember that a step is meant to represent a basic stream processing operation performed by a call to Kafka Streams. We may want to evolve how we perform a given operation in a way that is incompatible across versions. If a developer needs to modify a step in an incompatible way, they can do so by writing a new step type and advancing the version.
+
+### Longer Term
+The above proposal should cover the existing gaps in checking engine compatibility across versions. However it doesn't address many of the issues identified in the Background/Motivation section of this document:
+
+- Tests are always incomplete. There are always going to be gaps in what the tests cover, leaving us vulnerable to queries that are not verified for compatibility.
+- The tests take a long time to run. Currently there are 473 test cases that must be repeated for every previous version of KSQL. Today this takes about a minute. This is going to grow as we add language features and support CCloud, where we want to be able to release much more frequently.
+- The usage of compatibility-breaking configs adds a lot of complexity to the codebase. We have to maintain ugly code in the engine that switches on these configs to preserve legacy behaviour.
+
+To solve the above I propose that when accepting statements, in addition to persisting the original statements and configs, we persist the Execution Plan as well. Then, when executing queries, we can build the final streams topologies from the persisted Execution Plan instead of rebuilding the queries from statements every time. This way, we don't need compatibility verification, and don't need to worry about query compatibility in most of the engine code.
+
+This effectively splits the current engine into two layers:
+
+The upper layer builds Execution Plans, and consists of:
+
+- DDL evaluation
+- Parser
+- Analyzer
+- Logical Plan Builder
+- Physical Plan Builder
+
+Components in the upper layer would no longer need to worry about query compatibility.
```

Comment:
(Perhaps this is what you intended) I think the output of the PhyscialPlanBuilder should be an ExecutionPlan, and we pass that (and nothing else) into a new component which will turn that into the KStreams application. This ensures that we really have all of the information we need to re-create the exact query from the ExecutionPlan and we can be extra careful when evolving sub-components of that component.

Reply:
Yeah this is basically what I was getting at with the two layers. I think it would make sense to split them out into separate packages and have a strict interface between the two. Could definitely be more explicit/defined.